www/_posts/2018-08-01-future-cordova-ios-webview.md

```diff
@@ -0,0 +1,70 @@
+
+![Figure 1]({{ site.baseurl }}/static/img/blog/2018/arch-webview-engine.png)
+
+## Cordova iOS 5 with WKWebView
+
+Starting with Cordova iOS version 5, which is a planned upcoming release, we will ship both webview plugins with the cordova-ios platform, to enable developers to test and transition users to the new WKWebView component. Cordova iOS 5 will ship with a bridge webview plugin that can switch usage of the webview plugin used at runtime. Previously, you could only choose which webview you would use at build time.
+
+![Figure 2]({{ site.baseurl }}/static/img/blog/2018/bridge-webview-engine.png)
+
+<!--more-->
+
+## Migration and Transition using the Bridge WebView Plugin
+
+The developer can switch the webview used by setting a preference, that will be read at runtime when the app starts. Developers can transition users to the new WKWebView by user choice, in their app settings -- or randomly perhaps as an A/B test. This will give the developer and their users a chance to try out the new component, and they can back out and use UIWebView if there are any problems, without needing a new app release. This is intended to be used for testing and migration purposes only -- developers should aim to support only WKWebView functionality going forward.
+
+## Cordova iOS Future Release, WKWebView only
+
+When UIWebView support has been removed in a future iOS SDK, we will aim to release a future Cordova iOS version, which will remove UIWebView support, and WKWebView will then be the default webview engine.
+
+| Cordova iOS 4 >>> | Cordova iOS 5 >>> | Cordova iOS Future Release |
+|-------------------|-------------------|---------------|
+|  | UIWebView deprecated | UIWebView removed |
+|  | WKWebView Engine Plugin Integrated Into Platform | WKWebView Engine Plugin only |
+|  | Bridge WebView Plugin Added | Bridge WebView Plugin Removed |
+
+
+## Limitations of WKWebView
+
+There are [many limitations](https://mjtsai.com/blog/2018/06/20/webview-and-uiwebview-deprecated-in-favor-of-wkwebview/) of WKWebview, especially if you were using UIWebView previously. The limitations are:
+
+1. Cookies don't persist. This is a WebKit bug, but someone has
+created a plugin for a workaround. See
+[CB-12074](https://issues.apache.org/jira/browse/CB-12074)
+2. Can't delete cookies. This is/was a WebKit bug (2015), we need to test
+for the iOS 11/12. See [CB-11297](https://issues.apache.org/jira/browse/CB-11297)
+3. Can't execute JavaScript code in the background. There are several
+issues related to this. See
+[CB-12815](https://issues.apache.org/jira/browse/CB-12815)
+4. XmlHttpRequests don't work, because of Cross-Origin Resource
+Sharing issue (CORS). There is a workaround plugin created by Oracle
+(UPL licensed, which is Apache-2.0 compatible). See
+[CB-10143](https://issues.apache.org/jira/browse/CB-10143)
+5. Migration of localStorage from UIWebView. There is a migration
+plugin available. See [CB-11974](https://issues.apache.org/jira/browse/CB-11974)
+
```

Comment:
Should we add a note about iframes that (may) break?

Reply:
We could. I'll add something, but if you have a suggestion as well for the text, leave it here.

Reply:
Please add something like the following:

```
6. iframes will not be supported any longer, may be partially or completely broken
7. Known issues with WKWebView on iOS pre-11 which will be deprecated and dropped in a future Cordova release
```

(I remember discussion about something broken on WKWebView in iOS pre-11, forget exactly what it was)